### PR TITLE
Add sparse tensor mappings

### PR DESCRIPTION
**Pull request by karllessard**
Add minimal support for sparse tensors in TensorFlow Java. 

New factories like `TFloat32.sparseTensorOf(indices, values, denseShape)` allow users to get a typed reference to a sparse tensor so visit its data similarly to how it's done with dense tensors (with the exception that sparse tensors are read-only).

After this pull request, more will come to integrate sparse tensors within the TFJava ecosystem, including feeding sparse tensors to sessions/functions, creating constants, etc. The ultimate goal is that the user does not have to handle manually the individual dense tensors that compose a sparse one (i.e. the indices, the values and the dense shape) but right now as it is required, users can access them by casting back the returned reference to a `SparseTensor<T>`.

Requires https://github.com/tensorflow/java-ndarray/pull/5 to be merged first

CC\ @JimClarke5

### CODE REVIEW

**Review by Craigacp — CHANGES_REQUESTED** [review]

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * A virtual type of {@link Tensor} composed of three dense tensors (indices, values and dimensions) used to represent
+ * the sparse data into a multi-dimensional dense space.
+ *
+ * Any tensor returned by a sparse tensor factory (e.g. {@link TInt64#sparseTensorOf(TInt64, TInt64, TInt64)}) can be casted back
+ * to this interface to access directly the dense tensors it is composed of.
+ *
+ * @param <T> type of data stored in the tensor
+ */
+public interface SparseTensor<T extends TType> extends Tensor {
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java:15]: I had the same thought... but cannot find a good way to avoid. Ideally `Tensor` would also contain it's type but it is not possible since the type classes implement themselves the `Tensor` interface. :+1: for revisiting this later.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+  /**
+   * Creates a sparse tensor from {@code indices}, {@code values} and {@code denseShape} dense tensors.
+   *
+   * @param indices A 2-D tensor of shape {@code [N, ndims]}, that specifies the indices of the
+   *     elements in the sparse tensor that contain non-default values (elements are zero-indexed).
+   *     For example, {@code indices=[[1,3], [2,4]]} specifies that the elements with indexes of
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java:22]: I can change it but I took that example from the sparse NdArray factories, like [this one](https://github.com/tensorflow/java-ndarray/blob/d10262ebd79fd32f16035f25145bd1ac60a20fe4/ndarray/src/main/java/org/tensorflow/ndarray/NdArrays.java#L120), which seems to come in turn from TensorFlow python documentation for [sparse tensors](https://github.com/tensorflow/java-ndarray/blob/d10262ebd79fd32f16035f25145bd1ac60a20fe4/ndarray/src/main/java/org/tensorflow/ndarray/NdArrays.java#L120). So I guess if we are not happy with it, we should update the NdArray library as well?

CC\ @JimClarke5

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+   *     {@code [1,3]} and {@code [2,4]} have non-default values.
+   * @param values A 1-D tensor of shape {@code [N]}, which supplies the values for each
+   *     element in indices. For example, given {@code indices=[[1,3], [2,4]]}, the parameter {@code
+   *     values=[18, 3.8]} specifies that element {@code [1,3]} of the sparse tensor has a value of
+   *     {@code 18}, and element {@code [2,4]} of the tensor has a value of {@code 3.8}.
+   * @param denseShape A 1-D tensor of shape {@code [ndims]} where each the value at index {@code i}
+   *     represents to total number of element in dimension {@code i} in a dense version of that tensor.
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java:29]: Yes. Same comment as [here](https://github.com/tensorflow/java/pull/405#discussion_r790366649), if we update it in TensorFlow Core, we probably want to update it in the NdArray library as well.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      throw new IllegalArgumentException("Indices must have a coordinate for each dimensions of the tensor ["
+          + indices.shape().get(1) + " != " + denseShape.shape().get(0) + "]");
+    }
+    // Use mapper of the values tensor as this is the one giving the type of the sparse tensor as well
+    TensorMapper<T> mapper = (TensorMapper<T>) values.asRawTensor().typeInfo().mapper();
+    return mapper.mapSparse(indices, values, denseShape);
+  }
+
+  @Override
+  default RawTensor asRawTensor() {
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java:57]: I'll update the doc in the `Tensor` interface

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/types/TStringMapper.java [hunk 5]
```diff
@@ -100,4 +124,59 @@ public RawTensor asRawTensor() {
...
+      return new SparseTString(indices(), values().using(charset), denseShape());
+    }
+
+    @Override
+    public NdArray<byte[]> asBytes() {
+      return SparseNdArray.create(byte[].class, indices(), values().asBytes(), dimensions());
+    }
+
+    SparseTString(TInt64 indices, TString values, TInt64 denseShape) {
+      super(String.class, indices, values, null, SparseHelpers.toDimensionalSpace(denseShape));
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/types/TStringMapper.java:176]: Looking at the python's code, it should be an empty string. I'll fix it, thanks

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/types/TBfloat16Mapper.java [hunk 6]
```diff
@@ -55,4 +79,49 @@ public RawTensor asRawTensor() {
...
+    }
+
+    @Override
+    public long numBytes() {
+      return SparseHelpers.numBytes(this);
+    }
+
+    @Override
+    public void close() {
+      // sparse tensors do not own the dense tensors that compose them, nothing to close
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/types/TBfloat16Mapper.java:102]: Since all tensors created should normally be protected by a `try-with-resource` enclosure, they should normally be released after the sparse tensor, like in this snippet: 
```java
try (TInt64 indices = TInt64.tensorOf(...);
       TFloat64 values = TFloat64.vectorOf(...);
       TInt64 denseShape = TInt64.vectorOf(...);
       TFloat64 sparseTensor = TFloat64.sparseTensorOf(indices, values, denseShape)) {
...
}
```
But I agree that should be documented somewhere. I'm not super satisfied of this solution though and I'm now thinking that we might want to use JavaCPP references count instead, which would support more scenarios and be less error prone. Wdyt?

[the remark this answers is not in the document]
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/types/TBfloat16Mapper.java:102]: I think I'm fine with the current behaviour of the code, just it should be documented somewhere. I think reference counting is probably the way to go in the future, but I don't see a reason to hold this up while we do it.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java:29]: Ok, do you want to do the update to NdArray or shall I do it?

[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java:29]: I can do it :+1:

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/types/TBfloat16Mapper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/types/TBfloat16Mapper.java:102]: Actually, I've already got it working with reference counts. It still requires that all sub-components of a sparse tensor to be protected by `try-with-resources` block to avoid leaking, but it now allows the sparse tensor to live beyond that point. For example, we can support something like:
```java
TFloat64 createSparseTensor() {
    try (TInt64 indices = TInt64.tensorOf(...);
        TFloat64 values = TFloat64.vectorOf(...);
        TInt64 denseShape = TInt64.vectorOf(...)) {

        return TFloat64.sparseTensorOf(indices, values, denseShape);
    }
}

try (TFloat64 sparseTensor = createSparseTensor()) {
    ...
}
```
Should I go ahead with that implementation instead?

[the remark this answers is not in the document]
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/types/TBfloat16Mapper.java:102]: If you've already got it working, then yes that sounds better than the current approach.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    // Use mapper of the values tensor as this is the one giving the type of the sparse tensor as
+    // well
+    TensorMapper<T> mapper = (TensorMapper<T>) values.asRawTensor().typeInfo().mapper();
+
+    // Keep a strong reference to all sub-tensors of the sparse tensor
+    try (PointerScope scope = new PointerScope()) {
+      scope.attach(indices.asRawTensor().nativeHandle());
+      scope.attach(values.asRawTensor().nativeHandle());
+      scope.attach(denseShape.asRawTensor().nativeHandle());
+      return mapper.mapSparse(indices, values, denseShape, scope);
```
Comment (Craigacp) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java:105]: This only works because `mapSparse` calls extend on the scope right? In which case given this code is internal to our library, is it better to not use the scope in a try with resources and then remove the extend?

The fact that the scope looks like it's closed when you first read this method is quite confusing.

Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java:105]: I understand your concerns but this technique is used all over the place in our codebase (ex. [here](https://github.com/tensorflow/java/blob/1f691925187017346b19ec944a6ca6ebd83d2451/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/RawTensor.java#L149)). This is to prevent memory leaks in case something goes wrong before we can store a strong reference to the scope in an autocloseable object.

For example, if you attach the `indices` to the scope but fail to attach the `values` (let say that `values` has been closed before somehow), this will throw an exception but will leave the `indices` attached to the "phantom" scope that has not been closed properly and leave its reference count inaccurate. In such case, closing the `indices` later won't release it until that phantom scope is garbage collected (I think, not sure about this part).

In other words, this is extra safety we use everywhere in case something unexpected happens.

Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java:107]: done

**Review by Craigacp — APPROVED** [review]
